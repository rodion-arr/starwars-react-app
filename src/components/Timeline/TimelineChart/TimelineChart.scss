.timeline {
  --chart-col-width: 90px;
  --chart-first-col-width: 250px;
  --chart-row-height: 28px;

  & {
    padding: 1px var(--global-padding);
  }

  &__title {
    width: 100%;
    margin: 10px 0;
    padding: 5px 0;
    font-weight: bold;
    text-transform: uppercase;
    border-top: solid var(--main-color) 1px;
    border-bottom: solid var(--main-color) 1px;
  }

  &__chart-scroll {
    height: calc(100vh - 150px);
    width: 100%;
    overflow: scroll;
  }

  &__chart {
    display: block;
    width: max-content;
    table-layout: fixed;
    position: relative;
    border-collapse: collapse;

    &--header {
      position: sticky;
      top: 0;
      z-index: 1;
    }
  }

  &__chart-row {
    height: var(--chart-row-height);
    border-bottom: solid #505050 1px;

    & .timeline__chart-col:first-child {
      min-width: var(--chart-first-col-width);
      background: #1e1d1d;
      z-index: 1;
      box-shadow: rgb(149 157 165 / 20%) 0 8px 24px;
      position: sticky;
      left: 0;
    }

    &:hover .timeline__chart-col {
      background: #6d6d6d;
    }
  }

  &__chart-col {
    min-width: var(--chart-col-width);

    &--header {
      background: #1e1d1d;
      z-index: 1;
      box-shadow: rgb(149 157 165 / 20%) 0 8px 24px;
    }
  }

  &__chart-line {
    width: 100%;
    height: 10px;
    border-radius: 200px;
    background: white;
    box-sizing: border-box;
  }
}
