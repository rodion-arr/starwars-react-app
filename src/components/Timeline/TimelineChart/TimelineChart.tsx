import React, { useState, useEffect } from "react";
import {
  ChartRange,
  TimelineService,
} from "../../../services/timeline.service";
import { Timeline } from "../../../store/db/entities/timeline";
import "./TimelineChart.scss";

interface YearsList {
  title: string;
  rangeYear: number;
}

export interface Props {
  range: ChartRange;
  timeline: Record<string, Timeline>;
}

export const TimelineChart = ({ range, timeline }: Props): JSX.Element => {
  const [yearsData, setYearsData] = useState<YearsList[]>([]);

  // calculate chart year title
  useEffect(() => {
    const years: YearsList[] = [];
    for (let i = range.min; i <= range.max; i++) {
      years.push({
        title: TimelineService.getRangeTitle(i),
        rangeYear: i,
      });
    }

    setYearsData(years);
  }, [range]);

  return (
    <div className="timeline">
      <h2 className="timeline__title">Star Wars timeline</h2>

      {yearsData.length && (
        <div className="timeline__chart-scroll">
          <table className="timeline__chart timeline__chart--header">
            <thead>
              <tr className="timeline__chart-row">
                <th className="timeline__chart-col timeline__chart-col--header">
                  #
                </th>
                {yearsData.map((year) => (
                  <th
                    className="timeline__chart-col timeline__chart-col--header"
                    key={year.rangeYear}
                  >
                    {year.title}
                  </th>
                ))}
              </tr>
            </thead>
          </table>
          <table className="timeline__chart timeline__chart--body">
            <tbody>
              {Object.keys(timeline).map((entryId) => {
                const { start, end } = timeline[entryId];
                const delta = Math.abs(end - start) || 1;

                return (
                  <tr className="timeline__chart-row" key={entryId}>
                    <td className="timeline__chart-col">
                      {timeline[entryId].title}
                    </td>
                    {yearsData.map(({ rangeYear }) => {
                      let td = (
                        <td
                          className="timeline__chart-col"
                          key={rangeYear}
                        ></td>
                      );

                      if (start === rangeYear) {
                        td = (
                          <td
                            className="timeline__chart-col"
                            key={rangeYear}
                            colSpan={delta}
                          >
                            <div className="timeline__chart-line" />
                          </td>
                        );
                      }
                      return td;
                    })}
                  </tr>
                );
              })}
            </tbody>
          </table>
        </div>
      )}
    </div>
  );
};
