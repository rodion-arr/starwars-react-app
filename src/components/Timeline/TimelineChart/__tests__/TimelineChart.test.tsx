import { render } from "@testing-library/react";
import { TimelineChart } from "../TimelineChart";

describe("TimelineChart", () => {
  it("should render correctly", function () {
    const { container } = render(
      <TimelineChart
        range={{
          min: -10,
          max: 10,
        }}
        timeline={{
          "0": {
            id: "0",
            title: "The Phantom Menace",
            start: -9,
            end: -9,
          },
        }}
      />
    );

    expect(container).toMatchSnapshot();
  });
});
