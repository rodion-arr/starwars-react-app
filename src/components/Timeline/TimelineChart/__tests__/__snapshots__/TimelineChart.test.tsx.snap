// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimelineChart should have collapse functionality 1`] = `
<div>
  <div
    class="timeline timeline--collapsed"
  >
    <h2
      class="timeline__title"
    >
      Star Wars timeline
    </h2>
    <p>
      <strong>
        "Star Wars timeline"
      </strong>
       is a chart that visually represents the chronological order of Star Wars movies and series.
    </p>
    <p>
      The first column is movie/series name, the next columns are years in timeline at which this movie/series took place.
    </p>
    <h3>
      What do BBY and ABY mean?
    </h3>
    <p>
      In Star Wars galactic history the central event is considered as 
      <strong>
        Battle of Yavin
      </strong>
       (Episode IV).
    </p>
    <p>
      So BBY and ABY are "
      <strong>
        B
      </strong>
      efore 
      <strong>
        B
      </strong>
      attle of 
      <strong>
        Y
      </strong>
      avin" and "
      <strong>
        A
      </strong>
      fter 
      <strong>
        B
      </strong>
      attle of 
      <strong>
        Y
      </strong>
      avin" respectively. Or simply: before or after Eposide IV.
    </p>
    <div
      class="timeline__chart-scroll"
    >
      <table
        class="timeline__chart timeline__chart--header"
      >
        <thead>
          <tr
            class="timeline__chart-row"
          >
            <th
              class="timeline__chart-col timeline__chart-col--header timeline__chart-col--first-cell"
            >
              <span>
                #
              </span>
              <div
                class="timeline__collapse"
              />
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              10 BBY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              9 BBY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              8 BBY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              7 BBY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              6 BBY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              5 BBY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              4 BBY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              3 BBY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              2 BBY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              1 BBY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              0 BBY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              1 ABY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              2 ABY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              3 ABY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              4 ABY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              5 ABY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              6 ABY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              7 ABY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              8 ABY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              9 ABY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              10 ABY
            </th>
          </tr>
        </thead>
      </table>
      <table
        class="timeline__chart timeline__chart--body"
      >
        <tbody>
          <tr
            class="timeline__chart-row"
          >
            <td
              class="timeline__chart-col"
            >
              The Phantom Menace
            </td>
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
              colspan="1"
            >
              <div
                class="timeline__chart-line"
              />
            </td>
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`TimelineChart should render correctly 1`] = `
<div>
  <div
    class="timeline "
  >
    <h2
      class="timeline__title"
    >
      Star Wars timeline
    </h2>
    <p>
      <strong>
        "Star Wars timeline"
      </strong>
       is a chart that visually represents the chronological order of Star Wars movies and series.
    </p>
    <p>
      The first column is movie/series name, the next columns are years in timeline at which this movie/series took place.
    </p>
    <h3>
      What do BBY and ABY mean?
    </h3>
    <p>
      In Star Wars galactic history the central event is considered as 
      <strong>
        Battle of Yavin
      </strong>
       (Episode IV).
    </p>
    <p>
      So BBY and ABY are "
      <strong>
        B
      </strong>
      efore 
      <strong>
        B
      </strong>
      attle of 
      <strong>
        Y
      </strong>
      avin" and "
      <strong>
        A
      </strong>
      fter 
      <strong>
        B
      </strong>
      attle of 
      <strong>
        Y
      </strong>
      avin" respectively. Or simply: before or after Eposide IV.
    </p>
    <div
      class="timeline__chart-scroll"
    >
      <table
        class="timeline__chart timeline__chart--header"
      >
        <thead>
          <tr
            class="timeline__chart-row"
          >
            <th
              class="timeline__chart-col timeline__chart-col--header timeline__chart-col--first-cell"
            >
              <span>
                #
              </span>
              <div
                class="timeline__collapse"
              />
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              10 BBY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              9 BBY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              8 BBY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              7 BBY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              6 BBY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              5 BBY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              4 BBY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              3 BBY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              2 BBY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              1 BBY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              0 BBY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              1 ABY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              2 ABY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              3 ABY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              4 ABY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              5 ABY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              6 ABY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              7 ABY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              8 ABY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              9 ABY
            </th>
            <th
              class="timeline__chart-col timeline__chart-col--header"
            >
              10 ABY
            </th>
          </tr>
        </thead>
      </table>
      <table
        class="timeline__chart timeline__chart--body"
      >
        <tbody>
          <tr
            class="timeline__chart-row"
          >
            <td
              class="timeline__chart-col"
            >
              The Phantom Menace
            </td>
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
              colspan="1"
            >
              <div
                class="timeline__chart-line"
              />
            </td>
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
          </tr>
          <tr
            class="timeline__chart-row"
          >
            <td
              class="timeline__chart-col"
            >
              Episode 2
            </td>
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
            />
            <td
              class="timeline__chart-col"
              colspan="9"
            >
              <div
                class="timeline__chart-line"
              />
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
